rasp/tor_engine.py

```diff
@@ -1,57 +1,109 @@
+            can also be set with the ``RASP_TOR_CONTROL_PASSWORD``
+            environment variable.
+        control_password (str, optional): Password to protect control
+            port usage,
+            can also be set with the ``RASP_TOR_CONTROL_PORT``
+            environment variable.
+     """
+    def __init__(self,
+                 headers=None,
+                 address=None,
+                 port=None,
+                 control_port=None,
+                 control_password=None):
+        super(TorEngine, self).__init__(headers)
 
-        proxy_opener = urllib.request.build_opener(self.proxy_handler)
-        urllib.request.install_opener(proxy_opener)
+        self.address = (
+            address
+            or os.environ.get('RASP_TOR_ADDRESS')
+            or '127.0.0.1'
+        )
+        self.port = (
+            port
+            or os.environ.get('RASP_TOR_PORT')
+            or 9050
+        )
+        self.control_password = (
+            control_password
+            or os.environ.get('RASP_TOR_CONTROL_PASSWORD')
+            or getpass.getpass("Tor control password: ")
+        )
+        self.control_port = (
+            control_port
+            or os.environ.get('RASP_TOR_CONTROL_PORT')
+            or 9051
+        )
+
+        proxy_uri = 'socks5://{}:{}'.format(self.address, self.port)
+        proxies = {'http': proxy_uri, 'https': proxy_uri}
+        self.session.proxies.update(proxies)
 
     def __copy__(self):
         return TorEngine(
-            self.pw,
-            self.control,
-            self.signal,
-            self.proxy_handler,
             self.data,
-            self.headers
+            self.headers,
+            self.address,
+            self.port,
+            self.control_port,
+            self.control_password,
         )
 
-    def send_signal(self):
-        conn = stem.connection.connect(
-            control_port=self.control,
-            password=self.pw
-        )
-        conn.signal(self.signal)
-        conn.close()
-
-    def get_page_source(self, url):
-        try:
-            self.send_signal()
-            req = urllib.request.Request(url, self.data, self.headers)
-            res = urllib.request.urlopen(req)
-
-            try:
-                return Webpage(url, str(res.read()))
-            except http.client.IncompleteRead:
-                return
-        except urllib.error.HTTPError as e:
-            time.sleep(2)
+    def clean_circuits(self):
+        """Refreshes the mapping of nodes we connect through.
+
+        When a new path from the user to the destination is required, this
+        method refreshes the node path we use to connect. This gives us a new IP
+        address with which the destination sees."""
+        kwargs = {'address': self.address, 'port': self.control_port}
+        with Controller.from_port(**kwargs) as controller:
+            controller.authenticate(password=self.control_password)
```

Comment:
I could be wrong here, but I was pretty sure that `__del__` was essentially the python equivalent of the `tp_dealloc` member of `PyTypeObject`. I thought that `__exit__` and `__del__` were not exclusive so that `__del__` was essentially called after `__exit__`. Now that I think about it is implementing `__del__` when implementing `__exit__` an anti-pattern? I've never gone the context manager route before, so this is all new to me :-)

> make sure we don't cause a double free 

Example demonstrating what I was trying to say.

```python
import gc

class Test(object):
    def __init__(self):
        self.i = 0

    def info(self):
        return (self.i, hex(id(self)))

    def __enter__(self):
        self

    def __exit__(self, *a):
        self.i += 1
        print("__exit__: i={} &self={}".format(*self.info()))

    def __del__(self):
        self.i += 1
        print("__del__: i={} &self={}".format(*self.info()))

first = Test()
assert len(gc.get_referrers(first)) == 1
with first:
    assert len(gc.get_referrers(first)) == 2
assert len(gc.get_referrers(first)) == 1

with Test() as second:
    pass
```
Example run
```
$ python test.py
__exit__: i=1 &self=0x7eff20c76198
__exit__: i=1 &self=0x7eff20c76208
__del__: i=2 &self=0x7eff20c76208
__del__: i=2 &self=0x7eff20c76198
```